src/contracts/evm/Account/Account.ts

```diff
@@ -0,0 +1,62 @@
+import { NewAIF } from "types/contracts/evm";
+import { SimulContractTx, TxLog } from "types/evm";
+import { WalletState } from "contexts/WalletContext";
+import Contract from "../Contract";
+import abi from "./abi.json";
+
+/**
+ * By creating an event template constant and creating a type from it we enable the following:
+ *   1. The parent's event decoder (decodeEventInternal) knows how to populate the expected event arguments
+ *   2. The parent's event decoder can preserve type-safety
+ *   3. The Account can create a wrapper around the parent's event decoder to ensure only expected events can be decoded
+ *      using its (Account's) `decodeEvent` method
+ *
+ * IMPORTANT: it is crucial to have event arguments in the same order as they appear in the event from the logs.
+ */
+const eventTemplates = {
+  EndowmentCreated: {
+    endowId: 0,
+  },
+} as const;
+
+type Events = typeof eventTemplates;
+
+/**
+ * Representation of the `Account` contract that extends and wraps the generic `Contract` functionality.
+ * Ensures that `Contract` methods are called with the Account ABI.
+ */
+export default class Account extends Contract {
```

Comment:
@SovereignAndrey , @misicnenad - I have mentioned  in [pr comment](https://github.com/AngelProtocolFinance/angelprotocol-web-app/pull/1970#discussion_r1155815135) why this old approach is not to be copied moving forward to polygon, but since you still insist that this structure is to be retained and done initial works in parallel to #2025.  I would take the time here to explain in detail why I believe this approach is suboptimal considering how it fared throughout the app (juno setup),

long thread coming up.

Reply:
@misicnenad, @SovereignAndrey 

* ### Class structure (inheritance) works well with the nature of cosmos msgs, but not with the new EVM setup.

To encode a cosmos `msg` `MsgExecuteContract | MsgSend` or any other type of msg, the contract doesn't need to know what contract that msg came from.

that is
```ts
//all MsgExecuteContract
const msg1:MsgExecuteContract = account.createMsg(..args)
const msg2:MsgExecuteContract = multisig.createMsg(..args)
const msg3:MsgExecuteContract = registrar.createMsg(..args)

```
that's why currently in `contracts/Contract` there's an abstracted `createExecuteContractMsg`

```ts
createExecuteContractMsg(
    to: string,
    msg: object,
    funds: Coin[] = []
): MsgExecuteContractEncodeObject {
  return {
    typeUrl: "/cosmwasm.wasm.v1.MsgExecuteContract",
    value: {
      contract: to,
      sender: this.walletAddress,
      msg: toUtf8(JSON.stringify(msg)),
      funds,
    },
  };
}
```
which classes inheriting from `class Contracts` can use and just need to pass tx content.
```ts
//accounts.ts
createDepositMsg(payload: DepositPayload, funds: Coin[]) {
  return this.createExecuteContractMsg(
    Account.address,
    {
      deposit: payload,
    },
    funds
  );
}

```
this eventually will be encoded using `MsgExecuteContract.encode` method.


This is not the case with EVM contracts. The information to encode/decode transaction data is 
bound to the contract interface. Making a generic `Contract` and then extending it seems pointless

```ts
//Contract.ts
 protected constructor(
    abi: string[],
    contractAddress: string,
    wallet?: WalletState
  ) 

//Account.ts
constructor(contractAddress: string, wallet: WalletState) {
  super(abi, contractAddress, wallet);
}
```

`Account` has its own identity (`abi`), and has the capability to `encode/decode` transactions base on that identity. Why does it need to extend a generic `Contract`, pass that distinct identity and extend it later on?

You could just instantiate an account interface and use it directly in your tx and query definitions
```ts
//account/index.ts
const account = new Interface(account_abi)

//query.ts
({ id }) => accounts.encodeFunctionData("queryEndowmentDetails", [id])

//tx.ts
 "accounts.create-endowment": (aif) =>
    accounts.encodeFunctionData("createEndowment", [toTuple(aif)]),

```
no additional encode/decode wrappers


in case of events, just construct your event topic always on the context of the transaction where this topic is needed

```ts
const ev = accounts.getEvent("EndowmentCreated");
const topic = accounts.getEventTopic(ev);
/** 
 * NOTE use iface.encodeFilterTopics, to narrow down
 * topic (include specific arg values)
 * 
*/
```